torch_geometric/loader/dynamic_batch_sampler.py

```diff
@@ -0,0 +1,108 @@
+from typing import Iterator, List, Optional
+
+import torch
+
+from torch_geometric.data import Dataset
+
+
+class DynamicBatchSampler(torch.utils.data.sampler.Sampler[List[int]]):
+    r"""Dynamically adds samples to a mini-batch up to a maximum size (number of
+    nodes). When data samples have a wide range in sizes, specifying a
+    mini-batch size in terms of number of samples is not ideal and can cause
+    CUDA OOM errors.
+
+    Therefore, the number of steps per epoch is ambiguous, depending on the
+    order of the samples. By default the ``__len__()`` will be None, fine for
+    most cases but progress bars will be `infinite`. Alternatively,
+    ``num_steps`` can be supplied to cap the number of mini-batches produced
+    by the sampler.
+
+    **Usage:**
+
+    .. code-block:: python
+
+        from torch_geometric.loader import DataLoader, DynamicBatchSampler
+
+        batch_sampler = DynamicBatchSampler(dataset, max_num=10000)
+        loader = DataLoader(dataset, batch_sampler=sampler, ...)
+
+    Args:
+        data_source (Dataset): dataset to sample from, need this so that
+            potential samples can be sized to create the dynamic batch.
+        max_num (int): size of mini-batch to aim for in number of nodes or
+            edges.
+        mode (str, optional): `nodes` or `edges` to measure batch size.
+            (default: `nodes`)
+        shuffle (bool, optional): set to ``True`` to have the data reshuffled
+            at every epoch (default: ``False``).
+        skip_too_big (bool, optional): set to ``True`` to skip samples which
+            can't fit in a batch by itself. (default: ``False``).
+        num_steps (int, optional): The number of mini-batches to draw for a
+            single epoch. If set to :obj:`None`, will iterate through all the
+            underlying data, but __len__ will be :obj:`None` since it will be
+            ambiguous. (default: :obj:`None`)
+    """
+    def __init__(self, data_source: Dataset, max_num: int, mode: str = 'nodes',
+                 shuffle: bool = False, skip_too_big: bool = False,
+                 num_steps: Optional[int] = None):
+        if not isinstance(max_num, int) or max_num <= 0:
+            raise ValueError("`max_num` should be a positive integer value, "
+                             "but got max_num={max_num}.")
+        if mode not in ['nodes', 'edges']:
+            raise ValueError("`mode` choice should be either "
+                             f"`nodes` or `edges, not {mode}.")
+
+        self.max_num = max_num
+        self.data_source = data_source
+        self.shuffle = shuffle
+        self.skip_too_big = skip_too_big
+        if num_steps is None:
+            num_steps = len(data_source)
+        self.num_steps = num_steps
+
+        self.mode = f"num_{mode}"
+
+    def __iter__(self) -> Iterator[List[int]]:
+        batch = []
+        batch_n = 0
+        num_steps = 0
+        current_place = 0
```

Comment:
```suggestion
        num_processed = 0
```
I find `current_place` a bit misleading due to shuffle option.